src/Func.cpp

```diff
@@ -1124,11 +1124,26 @@ Stage Stage::specialize(Expr condition) {
             return Stage(specializations[i].definition, stage_name, dim_vars, storage_dims);
         }
     }
+
+    // Can't add any more specializations after specialize_fail().
+    for (size_t i = 0; i < specializations.size(); i++) {
+        user_assert(specializations[i].failure_message.empty()) << "Cannot add new specializations after specialize_fail().";
```

Comment:
If you can't add new specialization after specialize_fail, isn't that enough to check the last specialize in the list then?